### PR TITLE
feat(Nav): Convert Nav component to typescript, add Demos

### LINKED ISSUE #2018: Convert Nav to TypScript
#2005

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
**What**: 
-   Convert Nav component to typescript
-   Add Nav to Demo App
-   Add Nav integration tests

Fixes #2018

### CODE REVIEW

File: packages/patternfly-4/react-core/src/components/Nav/Nav.test.tsx [hunk 1]
```diff
@@ -20,8 +20,8 @@ beforeEach(() => {
 });
 
 const context = {
-  onSelect: () => undefined,
-  onToggle: () => undefined
+  onSelect: ():void => undefined,
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/Nav.test.tsx]: `() => undefined as any` is the pattern we're using so far.

File: packages/patternfly-4/react-core/src/components/Nav/Nav.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import { Omit } from '../../helpers/typeUtils';
+
+export interface NavProps extends Omit<React.DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement>, 'onSelect'> {
+  /** Anything that can be rendered inside of the nav */
+  children?: React.ReactNode;
+  /** Additional classes added to the container */
+  className?: string;
+  /** Callback for updating when item selection changes */
+  onSelect?(selectedItem: {groupId: number | string; itemId: number | string; to:string, event: React.FormEvent<HTMLInputElement>}): void;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/Nav.tsx]: Sorry to nitpick, but please change to `onSelect?: (selectedItem: {groupId: number | string; itemId: number | string; to:string, event: React.FormEvent<HTMLInputElement>}) => void;` (use arrow functions for the types).

File: packages/patternfly-4/react-core/src/components/Nav/Nav.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          {...props}
+        >
+          {children}
+        </nav>
+      </NavContext.Provider>
+    );
+  }
+}
+
+export default Nav;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/Nav.tsx]: Remove default exports from all files.

File: packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  /** Additional classes added to the container */
+  className?: string;
+  /** Group identifier, will be returned with the onToggle and onSelect callback passed to the Nav component */
+  groupId?: string | number;
+  /** If true makes the expandable list title active */
+  isActive?: boolean;
+  /** Identifier to use for the section aria label */
+  id?: string;
+  /** allow consumer to optionally override this callback and manage expand state externally */
+  onExpand?(e: React.MouseEvent<HTMLLIElement, MouseEvent>, val: boolean): void;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx]: Change to arrow function

File: packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+              <ul className={css(styles.navSimpleList)}>{children}</ul>
+            </section>
+          </li>
+        )}
+      </NavContext.Consumer>
+    );
+  }
+}
+
+export default NavExpandable;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx]: Remove default export.

File: packages/patternfly-4/react-core/src/components/Nav/NavGroup.tsx [hunk 6]
```diff
@@ -1,29 +1,27 @@
...
-  id: ''
-};
+export class NavGroup extends React.Component<NavGroupProps> {
+  static defaultProps = {
+    className: '',
+    id: ''
+  };
 
-class NavGroup extends React.Component {
+  static componentType = 'NavGroup';
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavGroup.tsx]: What's this needed for?

File: packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  /** Flag indicating whether the item is active */
+  isActive?: boolean;
+  /** Group identifier, will be returned with the onToggle and onSelect callback passed to the Nav component */
+  groupId?: string | number | null;
+  /** Item identifier, will be returned with the onToggle and onSelect callback passed to the Nav component */
+  itemId?: string | number | null;
+  /** If true prevents the default anchor link action to occur. Set to true if you want to handle navigation yourself. */
+  preventDefault?: boolean;
+  /** Callback for item click */
+  onClick: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>, itemId: number | string, groupId: number | string) => void
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx]: You used arrow syntax here :tada:

File: packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
+import * as React from 'react';
+import styles from '@patternfly/patternfly/components/Nav/nav.css';
+import { css } from '@patternfly/react-styles';
+import { Omit } from '../../helpers/typeUtils';
+import { NavContext } from './Nav';
+
+export interface NavItemProps extends Omit<React.HTMLProps<HTMLAnchorElement>,  'onClick'> {
+  /** Content rendered inside the nav item */
+  children?: React.ReactNode | string | null;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx]: I think `React.ReactNode` covers the `| string | null`. Does removing it causes a problem?

File: packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+  static defaultProps = {
+    children: null as React.ReactNode,
+    className: '',
+    to: '',
+    isActive: false,
+    groupId: null as string,
+    itemId: null as string,
+    preventDefault: false,
+    onClick: null as ((event: React.MouseEvent<HTMLAnchorElement, MouseEvent>, itemId: number | string, groupId: number | string) => void)
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx]: Create a type for this long thing and reference it in the props and here.

File: packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  };
+
+  static componentType:string = 'NavItem';
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavItem.tsx]: What's this for?

File: packages/patternfly-4/react-core/src/components/Nav/NavItemSeparator.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import * as React from 'react';
+import styles from '@patternfly/patternfly/components/Nav/nav.css';
+import { css } from '@patternfly/react-styles';
+
+export class NavItemSeparator extends React.Component<React.HTMLProps<HTMLLIElement>> {
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavItemSeparator.tsx]: Can we make this back into a function component?

File: packages/patternfly-4/react-core/src/components/Nav/NavList.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+import {OneOf} from '../../helpers/typeUtils';
+
+export interface NavListProps extends React.DetailedHTMLProps<React.HTMLAttributes<HTMLUListElement>, HTMLUListElement> {
+  /** Children nodes */
+  children?: React.ReactNode;
+  /** Additional classes added to the list */
+  className?: string;
+  /** Indicates the list type. */
+  variant?: OneOf<typeof NavVariants, keyof typeof NavVariants>;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavList.tsx]: I think just `NavVariants` will do. If not, use `NavVariants | keyof typeof NavVariants`.

File: packages/patternfly-4/react-core/src/components/Nav/NavList.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+export class NavList extends React.Component<NavListProps> {
+
+  static defaultProps = {
+    variant: 'default',
+    children: null as React.ReactNode,
+    className: ''
+  };
+
+  static componentType = 'NavList';
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/NavList.tsx]: What's this do?

File: packages/patternfly-4/react-core/src/declarations.d.ts [hunk 14]
```diff
@@ -3,4 +3,18 @@
 declare module '@patternfly/patternfly/components/*.css' {
   const styles: any;
   export default styles;
-}
\ No newline at end of file
+}
+
+// We use a JS babel plugin to resolve these imports that `tsc` doesn't
+// know about, so pretend that everything will work
+declare module '@patternfly/patternfly/layouts/*.css' {
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/declarations.d.ts]: I like this solution, but I think my PR is about to be merged (Alert) that just changes it to `declare module '@patternfly/patternfly/*/*.css'`, so you might want to go with that or deal with merge conflicts later.

File: packages/patternfly-4/react-integration/cypress/integration/nav.spec.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    });
+  });
+
+  it('Verify Default Nav', () => {
+    cy.get('#nav-primary-default .pf-c-nav__link').each((defaultNavLink: JQuery<HTMLAnchorElement>, index: number) => {
+      const isCurrent = defaultNavLink.hasClass('pf-m-current');
+      expect(isCurrent).to.be.equal(index === 0);
+    });
+    cy.get('#default-link3').then((defaultLink3: JQuery<HTMLAnchorElement>) => {
+      cy.wrap(defaultLink3).click();
```
Comment (redallen) [inline packages/patternfly-4/react-integration/cypress/integration/nav.spec.ts]: Wow, thanks for the awesome tests!

**Comment by jeff-phillips-18** [comment]
> Awesome work! Can we use `git mv` to rename the primary component files instead of deleting and recreating them? If we can do this, we'll retain the ability to trace the history of these modules going forward which is important for tracking down when certain changes were made, by who, and why. Looks like you've already done this for several components.

I did that. I tried my darnedest to keep then from being delete/add's but somehow they all got borked along the way (think it was a `git commit --amend`).

**Comment by redallen** [comment]
@jeff-phillips-18 the way I've been doing it is to reset to the commit before your initial commit on this branch. So `git reset --soft ca245d37e856fd7dba9b364c26329af001cddd98~`. Then move your new .tsx files somewhere temporarily. Rename each .js -> .tsx. Commit. In the next commit change the contents of the tsx. It's a pain and requires two commits, but it keeps the history.

**Comment by jeff-phillips-18** [comment]
@redallen Have you noticed if it stays as a rename after a `Squash and merge`?

**Comment by redallen** [comment]
No idea, I guess we'll find out how Github does that part when the first TS PR gets merged. That's actually a really good question.

**Comment by seanforyou23** [comment]
> Do you want me to try again?

No worries @jeff-phillips-18 - just where you can without getting blocked. I know we're converting a lot of components over the next while so more than anything I wanted to throw it out there so we can save as much history as possible and adopt this process where we can going forward. I should have gotten out in front of this before all these PR's popped up. That said, @redallen 's approach above should work. 

As far as squash and merge, I think if step zero of the working branch is to rename the file with git mv, then the subsequent commits operate on those files, logs should reflect the full history just fine. Thanks for looking into this guys, it's a big win for people inside and outside the project.

**Comment by jeff-phillips-18** [comment]
@redallen @seanforyou23 Updates made based on comments

File: packages/patternfly-4/react-core/src/components/Nav/Nav.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  onToggle(event: React.MouseEvent<HTMLInputElement>, groupId: number | string, toggleValue: boolean) {
+    this.props.onToggle({
+      event,
+      groupId,
+      isExpanded: toggleValue
+    });
+  }
+
+  render() {
+    const { 'aria-label': ariaLabel, children, className, onSelect, ...props } = this.props;
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/components/Nav/Nav.tsx]: i see you added onSelect here, maybe also spread out onToggle?

File: packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+export interface NavExpandableProps extends React.DetailedHTMLProps<React.LiHTMLAttributes<HTMLLIElement>, HTMLLIElement> {
+  /** Title shown for the expandable list */
+  title: string;
+  /** If defined, screen readers will read this text instead of the list title */
+  srText?: string;
+  /** Boolean to programatically expand or collapse section */
+  isExpanded?: boolean;
+  /** Anything that can be rendered inside of the expandable list */
+  children?: React.ReactNode | string;
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/components/Nav/NavExpandable.tsx]: ReactNode should contain string, so this could just be ReactNode

**Comment by jschuler** [comment]
I noticed that the page layout demos did not add preventDefault to the NavItem components. Adding it fixes the demo issues
https://github.com/jeff-phillips-18/patternfly-react/pull/1

**Comment by jeff-phillips-18** [comment]
Yes, the issue did exist prior to this PR. I added a change to auto-`preventDefault` for default nav item links if no `to` property is passed. Removed the unwanted `to` settings from the demo pages.

File: packages/patternfly-4/react-core/src/components/Nav/Nav.tsx [hunk 18]
```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import styles from '@patternfly/patternfly/components/Nav/nav.css';
+import { css } from '@patternfly/react-styles';
+
+import { Omit } from '../../helpers/typeUtils';
+
+export type NavSelectClickHandler = (
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Nav/Nav.tsx]: Woohoo, this is readable!

**Comment by jeff-phillips-18** [comment]
@jschuler I believe all updates have been made

**Pull request merged**